File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] prefix = new int[nums.length];
        int[] suffix = new int[nums.length];
        prefix[0] = 1;
        suffix[nums.length - 1] = 1;
        int i = 1;
        int curr = 1;
        while(i < nums.length){
            curr = curr * nums[i-1];
            System.out.println(curr);
            prefix[i] = curr;
            i++;
        }
        i = nums.length - 2;
        curr = 1;
        while (i >= 0){
            curr = curr * nums[i+1];
            System.out.println(curr);
            suffix[i] = curr;
            i--;
        }
        i = 0;
        while(i < nums.length){
            prefix[i] = prefix[i] * suffix[i];
            i++;
        }
        return prefix;

    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1] * len(nums)
        suffix = [1] * len(nums)
        result = [1] * len(nums)
        i = 1
        while i < len(nums):
            prefix[i] = prefix[i-1] * nums[i-1]
            i = i + 1
        i = len(nums) - 2
        while i >= 0:
            suffix[i] = suffix[i + 1] * nums[i + 1]
            i = i - 1
        i = 0
        while i < len(nums):
            result[i] = prefix[i] * suffix[i]
            i = i + 1
        return result
        
